### PR TITLE
Override memory-based Read/WriteAsync APIs on System.IO streams

### PR DESCRIPTION
**Pull request by stephentoub**
BufferedStream, PipeStream, DeflateStream, and GZipStream

Fixes https://github.com/dotnet/corefx/issues/22397
Fixes https://github.com/dotnet/corefx/issues/22392
Fixes https://github.com/dotnet/corefx/issues/22390
Fixes https://github.com/dotnet/corefx/issues/22395

cc: @JeremyKuhne, @ianhays, @KrzysztofCwalina, @ahsonkhan

### CODE REVIEW

File: src/System.IO.Compression/tests/DeflateStreamTests.cs [hunk 1]
```diff
@@ -799,7 +827,7 @@ public void Roundtrip_Write_ReadByte()
         public async Task WrapNullReturningTasksStream()
         {
             using (var ds = new DeflateStream(new BadWrappedStream(BadWrappedStream.Mode.ReturnNullTasks), CompressionMode.Decompress))
-                await Assert.ThrowsAsync<InvalidOperationException>(() => ds.ReadAsync(new byte[1024], 0, 1024));
+                await Assert.ThrowsAsync<ArgumentNullException>(() => ds.ReadAsync(new byte[1024], 0, 1024));
```
Comment (ahsonkhan) [inline src/System.IO.Compression/tests/DeflateStreamTests.cs]: What caused the need for this change in exception thrown?

**Comment by danmosemsft** [comment]
@dotnet-bot help please (piggy back)

File: src/System.IO.Compression/tests/DeflateStreamTests.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.IO.Compression/tests/DeflateStreamTests.cs]: The issue is that I changed DeflateStream to use the new ReadAsync overload, which returns a ValueTask.  If the target stream doesn't override the new overload, the base implementation will delegate to the existing ReadAsync overload, which returns Task, and then construct a ValueTask from that.  This bad stream returns null from ReadAsync, causing the ValueTask ctor to fail.  Technically it's a breaking change, but the stream is already in a terribly broken state if it's returning null from ReadAsync, and so which exception emerges in that case didn't seem like a reason to not use the new method.

Comment (ahsonkhan) [reply src/System.IO.Compression/tests/DeflateStreamTests.cs]: > Technically it's a breaking change, but the stream is already in a terribly broken state if it's returning null from ReadAsync, and so which exception emerges in that case didn't seem like a reason to not use the new method.

I don't know if these APIs have shipped/etc. (can you please give some background)... so, I am not sure if we care if this is a breaking change. I lack the context, but are there any concerns with this?

Comment (stephentoub) [reply src/System.IO.Compression/tests/DeflateStreamTests.cs]: > I don't know if these APIs have shipped

DeflateStream has been shipping for a long time.

> are there any concerns with this?

I'm not concerned.  But I welcome other's opinions... @weshaggard, @terrajobst, @JeremyKuhne?  The issue is that if someone has a Stream where they've overridden ReadAsync(byte[], ...) and are returning null from it, and don't override ReadAsync(Memory, ...), and then that Stream is wrapped in a DeflateStream, previously DeflateStream would throw an InvalidOperationException, now it ends up throwing an ArgumentNullException.  Do we care?

Comment (JeremyKuhne) [reply src/System.IO.Compression/tests/DeflateStreamTests.cs]: I'm not concerned either- I wouldn't bend over backwards for this.

**Pull request merged**